tests/all/component_model/call_hook.rs

```diff
@@ -0,0 +1,683 @@
+            assert_eq!(store.data().context.last(), Some(&Context::Host));
+            assert_eq!(store.data().calls_into_host, store.data().calls_into_wasm);
+
+            match action {
+                TRAP_IN_F => bail!("trapping in f"),
+                TRAP_NEXT_CALL_HOST => store.data_mut().trap_next_call_host = true,
+                TRAP_NEXT_RETURN_HOST => store.data_mut().trap_next_return_host = true,
+                TRAP_NEXT_CALL_WASM => store.data_mut().trap_next_call_wasm = true,
+                TRAP_NEXT_RETURN_WASM => store.data_mut().trap_next_return_wasm = true,
+                _ => {} // Do nothing
+            }
+
+            Ok(())
+        })?;
+
+    let wat = r#"
+        (component
+            (import "f" (func $f (param "action" s32)))
+
+            (core func $f_lower
+                (canon lower (func $f))
+            )
+            (core module $m
+                (import "" "" (func $f (param i32)))
+
+                (func $export (param i32)
+                    (call $f (local.get 0))
+                )
+
+                (export "export" (func $export))
+            )
+            (core instance $i (instantiate $m
+                (with "" (instance
+                    (export "" (func $f_lower))
+                ))
+            ))
+            (func (export "export") (param "action" s32)
+                (canon lift
+                    (core func $i "export")
+                )
+            )
+        )
+    "#;
+
+    let component = Component::new(&engine, wat)?;
+
+    let run = |action: i32, again: bool| -> (State, Option<anyhow::Error>) {
+        let mut store = Store::new(&engine, State::default());
+        store.call_hook(sync_call_hook);
+        let inst = linker
+            .instantiate(&mut store, &component)
+            .expect("instantiate");
+
+        let export = inst
+            .get_typed_func::<(i32,), ()>(&mut store, "export")
+            .expect("looking up `export`");
+
+        let mut r = export.call(&mut store, (action,));
+        if r.is_ok() && again {
+            export.post_return(&mut store).unwrap();
+            r = export.call(&mut store, (action,));
+        }
+        (store.into_data(), r.err())
+    };
+
+    let (s, e) = run(DO_NOTHING, false);
+    assert!(e.is_none());
+    assert_eq!(s.calls_into_host, 1);
+    assert_eq!(s.returns_from_host, 1);
+    assert_eq!(s.calls_into_wasm, 1);
+    assert_eq!(s.returns_from_wasm, 1);
+
+    let (s, e) = run(DO_NOTHING, true);
+    assert!(e.is_none(), "{:?}", e);
+    assert_eq!(s.calls_into_host, 2);
+    assert_eq!(s.returns_from_host, 2);
+    assert_eq!(s.calls_into_wasm, 2);
+    assert_eq!(s.returns_from_wasm, 2);
+
+    let (s, e) = run(TRAP_IN_F, false);
+    assert!(format!("{:?}", e.unwrap()).contains("trapping in f"));
+    assert_eq!(s.calls_into_host, 1);
+    assert_eq!(s.returns_from_host, 0);
+    assert_eq!(s.calls_into_wasm, 1);
+    assert_eq!(s.returns_from_wasm, 0);
```

Comment:
This differs from the core wasm behavior, in that trapping in `f` does not register the return from host or return from wasm hooks. @alexcrichton, is this reasonable, or is there something missing from this PR?